src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java

```diff
@@ -367,24 +390,127 @@ private static String getClassNameFromQualifiedName(String qualifiedName) {
         return qualifiedName.substring(qualifiedName.lastIndexOf(PACKAGE_SEPARATOR) + 1);
     }
 
-    /** Maintains information about class' ctors. */
-    private static final class ClassDesc {
+    /**
+     * Calculates and returns the type declaration matching count when {@code classToBeMatched} is
+     * considered to be super class of an anonymous inner class.
+     *
+     * <p>
+     * Suppose our pattern class is {@code Main.ClassOne} and class to be matched is
+     * {@code Main.ClassOne.ClassTwo.ClassThree} then type declaration name matching count would
+     * be calculated by comparing every character, and updating main counter when we hit "." or
+     * when it is the last character of the pattern class and certain conditions are met. This is
+     * done so that matching count is 13 instead of 5. This is due to the fact that pattern class
+     * can contain anonymous inner class object of a nested class which isn't true in case of
+     * extending classes as you can't extend nested classes.
+     * </p>
+     *
+     * @param patternTypeDeclaration type declaration against which the given type declaration has
+     *                               to be matched
+     * @param typeDeclarationToBeMatched type declaration to be matched
+     * @return type declaration matching count
+     */
+    private static int getAnonSuperTypeMatchingCount(String patternTypeDeclaration,
+                                                    String typeDeclarationToBeMatched) {
+        final int typeDeclarationToBeMatchedLength = typeDeclarationToBeMatched.length();
+        final int minLength = Math
+            .min(typeDeclarationToBeMatchedLength, patternTypeDeclaration.length());
+        final char packageSeparator = PACKAGE_SEPARATOR.charAt(0);
+        final boolean shouldCountBeUpdatedAtLastCharacter =
+            typeDeclarationToBeMatchedLength > minLength
+                && typeDeclarationToBeMatched.charAt(minLength) == packageSeparator;
+
+        int result = 0;
+        for (int index = 0; index < minLength && patternTypeDeclaration.charAt(
+            index) == typeDeclarationToBeMatched.charAt(index); ++index) {
+            if (index == minLength - 1
+                && shouldCountBeUpdatedAtLastCharacter
+                || patternTypeDeclaration.charAt(index) == packageSeparator) {
+                result = index;
+            }
+        }
```

Comment:
Please do:

```suggestion
        for (int index = 0; 
                index < minLength 
                    && patternTypeDeclaration.charAt(index) == typeDeclarationToBeMatched.charAt(index); 
                index++) {
                
            if (index == minLength - 1 && shouldCountBeUpdatedAtLastCharacter
                    || patternTypeDeclaration.charAt(index) == packageSeparator) {
                result = index;
            }
        }
```

To make this section easier to read by humans.

Reply:
Done, had to change indent to +2 to adhere to line length.

Reply:
Resolved.